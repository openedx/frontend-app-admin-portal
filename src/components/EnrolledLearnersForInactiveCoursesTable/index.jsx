import React from 'react';

import { useIntl } from '@edx/frontend-platform/i18n';

import TableContainer from '../../containers/TableContainer';
import { i18nFormatTimestamp } from '../../utils';
import EnterpriseDataApiService from '../../data/services/EnterpriseDataApiService';

const EnrolledLearnersForInactiveCoursesTable = () => {
  const intl = useIntl();

  const tableColumns = [
    {
      label: intl.formatMessage({
        id: 'admin.portal.lpr.enrolled.learners.inactive.courses.table.user_email.column.heading',
        defaultMessage: 'Email',
        description: 'Column heading for the user email column in the enrolled learners table for inactive courses',
      }),
      key: 'user_email',
      columnSortable: true,
    },
    {
      label: intl.formatMessage({
        id: 'admin.portal.lpr.enrolled.learners.inactive.courses.table.enrollment_count.column.heading',
        defaultMessage: 'Total Course Enrollment Count',
        description: 'Column heading for the course enrollment count column in the enrolled learners table for inactive courses',
      }),
      key: 'enrollment_count',
      columnSortable: true,
    },
    {
      label: intl.formatMessage({
        id: 'admin.portal.lpr.enrolled.learners.inactive.courses.table.course_completion_count.column.heading',
        defaultMessage: 'Total Completed Courses Count',
        description: 'Column heading for the completed courses count column in the enrolled learners table for inactive courses',
      }),
      key: 'course_completion_count',
      columnSortable: true,
    },
    {
      label: intl.formatMessage({
        id: 'admin.portal.lpr.enrolled.learners.inactive.courses.table.last_activity_date.column.heading',
        defaultMessage: 'Last Activity Date',
        description: 'Column heading for the last activity date column in the enrolled learners table for inactive courses',
      }),
      key: 'last_activity_date',
      columnSortable: true,
    },
  ];

  const formatLearnerData = learners => learners.map(learner => ({
    ...learner,
    user_email: <span data-hj-suppress>{learner.user_email}</span>,
    last_activity_date: i18nFormatTimestamp({
      intl, timestamp: learner.last_activity_date,
    }),
  }));

  return (
    <TableContainer
      id="enrolled-learners-inactive-courses"
      className="enrolled-learners-inactive-courses"
      fetchMethod={EnterpriseDataApiService.fetchEnrolledLearnersForInactiveCourses}
      columns={tableColumns}
      formatData={formatLearnerData}
      tableSortable
    />
  );
};

export default EnrolledLearnersForInactiveCoursesTable;
