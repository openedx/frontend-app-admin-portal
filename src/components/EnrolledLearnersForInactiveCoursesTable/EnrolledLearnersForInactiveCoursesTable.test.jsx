import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

import EnrolledLearnersForInactiveCoursesTable from '.';

const enterpriseId = 'test-enterprise';
const mockStore = configureMockStore([thunk]);
const enrolledLearnersForInactiveCoursesEmptyStore = mockStore({
  portalConfiguration: {
    enterpriseId,
  },
  table: {
    'enrolled-learners-inactive-courses': {
      data: {
        results: [],
        current_page: 1,
        num_pages: 1,
      },
      ordering: null,
      loading: false,
      error: null,
    },
  },
});
const enrolledLearnersForInactiveCoursesStore = mockStore({
  portalConfiguration: {
    enterpriseId,
  },
  table: {
    'enrolled-learners-inactive-courses': {
      data: {
        count: 3,
        num_pages: 1,
        current_page: 1,
        results: [
          {
            id: 1,
            enterprise_id: '72416e52-8c77-4860-9584-15e5b06220fb',
            lms_user_id: 11,
            enterprise_user_id: 222,
            enterprise_sso_uid: 'harry',
            user_account_creation_timestamp: '2015-02-12T23:14:35Z',
            user_email: 'test_user_1@example.com',
            user_username: 'test_user_1',
            user_country_code: 'US',
            last_activity_date: '2017-06-23',
            enrollment_count: 2,
            course_completion_count: 1,
          },
          {
            id: 1,
            enterprise_id: '72416e52-8c77-4860-9584-15e5b06220fb',
            lms_user_id: 22,
            enterprise_user_id: 333,
            enterprise_sso_uid: 'harry',
            user_account_creation_timestamp: '2016-05-12T22:14:36Z',
            user_email: 'test_user_2@example.com',
            user_username: 'test_user_2',
            user_country_code: 'US',
            last_activity_date: '2018-01-15',
            enrollment_count: 5,
            course_completion_count: 5,
          },
          {
            id: 1,
            enterprise_id: '72416e52-8c77-4860-9584-15e5b06220fb',
            lms_user_id: 33,
            enterprise_user_id: 444,
            enterprise_sso_uid: 'harry',
            user_account_creation_timestamp: '2017-12-12T18:10:15Z',
            user_email: 'test_user_3@example.com',
            user_username: 'test_user_3',
            user_country_code: 'US',
            last_activity_date: '2017-11-18',
            enrollment_count: 6,
            course_completion_count: 4,
          },
        ],
        next: null,
        start: 0,
        previous: null,
      },
      ordering: null,
      loading: false,
      error: null,
    },
  },
});

const EnrolledLearnersForInactiveCoursesEmptyTableWrapper = props => (
  <MemoryRouter>
    <Provider store={enrolledLearnersForInactiveCoursesEmptyStore}>
      <EnrolledLearnersForInactiveCoursesTable
        {...props}
      />
    </Provider>
  </MemoryRouter>
);

const EnrolledLearnersForInactiveCoursesWrapper = props => (
  <MemoryRouter>
    <Provider store={enrolledLearnersForInactiveCoursesStore}>
      <EnrolledLearnersForInactiveCoursesTable
        {...props}
      />
    </Provider>
  </MemoryRouter>
);

describe('EnrolledLearnersForInactiveCoursesTable', () => {
  it('renders empty state correctly', () => {
    const { container: tree } = render(
      <EnrolledLearnersForInactiveCoursesEmptyTableWrapper />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders enrolled learners for inactive courses table correctly', () => {
    const { container: tree } = render(
      <EnrolledLearnersForInactiveCoursesWrapper />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('renders enrolled learners for inactive courses table with correct data', () => {
    const tableId = 'enrolled-learners-inactive-courses';
    const columnTitles = [
      'Email', 'Total Course Enrollment Count', 'Total Completed Courses Count', 'Last Activity Date',
    ];
    const rowsData = [
      [
        'test_user_1@example.com',
        '2',
        '1',
        'June 23, 2017',
      ],
      [
        'test_user_2@example.com',
        '5',
        '5',
        'January 15, 2018',
      ],
      [
        'test_user_3@example.com',
        '6',
        '4',
        'November 18, 2017',
      ],
    ];

    const wrapper = render((
      <EnrolledLearnersForInactiveCoursesWrapper />
    ));

    // Verify that table has correct number of columns
    expect(wrapper.container.querySelectorAll(`.${tableId} thead th`).length).toEqual(columnTitles.length);

    // Verify only expected columns are shown
    wrapper.container.querySelectorAll(`.${tableId} thead th`).forEach((column, index) => {
      expect(column.textContent).toContain(columnTitles[index]);
    });

    // Verify that table has correct number of rows
    expect(wrapper.container.querySelectorAll(`.${tableId} tbody tr`).length).toEqual(rowsData.length);

    // Verify each row in table has correct data
    wrapper.container.querySelectorAll(`.${tableId} tbody tr`).forEach((row, rowIndex) => {
      row.querySelectorAll('td').forEach((cell, colIndex) => {
        expect(cell.textContent).toEqual(rowsData[rowIndex][colIndex]);
      });
    });
  });
});
