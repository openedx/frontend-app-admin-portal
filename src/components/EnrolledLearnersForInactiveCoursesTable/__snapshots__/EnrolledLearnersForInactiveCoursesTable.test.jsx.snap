// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnrolledLearnersForInactiveCoursesTable renders empty state correctly 1`] = `
<div
  className="fade alert-content alert alert-warning show"
  role="alert"
>
  <span
    className="pgn__icon alert-icon"
  >
    <svg
      aria-hidden={true}
      fill="none"
      focusable={false}
      height={24}
      role="img"
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
        fill="currentColor"
      />
    </svg>
  </span>
  <div
    className="pgn__alert-message-wrapper"
  >
    <div
      className="alert-message-content"
    >
      There are no results.
    </div>
  </div>
</div>
`;

exports[`EnrolledLearnersForInactiveCoursesTable renders enrolled learners for inactive courses table correctly 1`] = `
<div
  className="enrolled-learners-inactive-courses"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="table-responsive"
      >
        <table
          className="table table-sm table-striped"
        >
          <thead
            className=""
          >
            <tr
              className=""
            >
              <th
                className="sortable"
                scope="col"
              >
                <button
                  className="btn btn-header"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    Email
                    <span
                      className="sr-only"
                    >
                       
                      sort ascending
                    </span>
                     
                    <span
                      aria-hidden={true}
                      className="fa fa-sort-asc"
                    />
                  </span>
                </button>
              </th>
              <th
                className="sortable"
                scope="col"
              >
                <button
                  className="btn btn-header"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    Total Course Enrollment Count
                    <span
                      className="sr-only"
                    >
                       
                      click to sort
                    </span>
                     
                    <span
                      aria-hidden={true}
                      className="fa fa-sort"
                    />
                  </span>
                </button>
              </th>
              <th
                className="sortable"
                scope="col"
              >
                <button
                  className="btn btn-header"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    Total Completed Courses Count
                    <span
                      className="sr-only"
                    >
                       
                      click to sort
                    </span>
                     
                    <span
                      aria-hidden={true}
                      className="fa fa-sort"
                    />
                  </span>
                </button>
              </th>
              <th
                className="sortable"
                scope="col"
              >
                <button
                  className="btn btn-header"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    Last Activity Date
                    <span
                      className="sr-only"
                    >
                       
                      click to sort
                    </span>
                     
                    <span
                      aria-hidden={true}
                      className="fa fa-sort"
                    />
                  </span>
                </button>
              </th>
            </tr>
          </thead>
          <tbody
            className=""
          >
            <tr
              className=""
            >
              <td
                className=""
                scope={null}
              >
                <span
                  data-hj-suppress={true}
                >
                  test_user_1@example.com
                </span>
              </td>
              <td
                className=""
                scope={null}
              >
                2
              </td>
              <td
                className=""
                scope={null}
              >
                1
              </td>
              <td
                className=""
                scope={null}
              >
                June 23, 2017
              </td>
            </tr>
            <tr
              className=""
            >
              <td
                className=""
                scope={null}
              >
                <span
                  data-hj-suppress={true}
                >
                  test_user_2@example.com
                </span>
              </td>
              <td
                className=""
                scope={null}
              >
                5
              </td>
              <td
                className=""
                scope={null}
              >
                5
              </td>
              <td
                className=""
                scope={null}
              >
                January 15, 2018
              </td>
            </tr>
            <tr
              className=""
            >
              <td
                className=""
                scope={null}
              >
                <span
                  data-hj-suppress={true}
                >
                  test_user_3@example.com
                </span>
              </td>
              <td
                className=""
                scope={null}
              >
                6
              </td>
              <td
                className=""
                scope={null}
              >
                4
              </td>
              <td
                className=""
                scope={null}
              >
                November 18, 2017
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <div
    className="row mt-2"
  >
    <div
      className="col d-flex justify-content-center"
    >
      <nav
        aria-label="enrolled-learners-inactive-courses-pagination"
        className="pagination-default"
      >
        <div
          aria-atomic={true}
          aria-live="polite"
          aria-relevant="text"
          className="sr-only"
        >
          Page 1, Current Page, of 1
        </div>
        <ul
          className="pagination"
        >
          <li
            className="page-item disabled"
          >
            <button
              aria-label="Previous"
              className="previous page-link btn btn-primary"
              disabled={true}
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              <div>
                <span
                  className="pgn__icon"
                >
                  <svg
                    aria-hidden={true}
                    fill="none"
                    focusable={false}
                    height={24}
                    role="img"
                    viewBox="0 0 24 24"
                    width={24}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12l4.58-4.59z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                Previous
              </div>
            </button>
          </li>
          <li
            className="page-item disabled"
          >
            <button
              aria-label="Next"
              className="next page-link btn btn-primary"
              disabled={true}
              onClick={[Function]}
              tabIndex="-1"
              type="button"
            >
              <div>
                Next
                <span
                  className="pgn__icon"
                >
                  <svg
                    aria-hidden={true}
                    fill="none"
                    focusable={false}
                    height={24}
                    role="img"
                    viewBox="0 0 24 24"
                    width={24}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6-6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </div>
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</div>
`;
